File: deepseek-harness/go/permissions.go
package dsagent

import (
	"bufio"
	"fmt"
	"io"
	"strings"
)

// Mode controls when a mutating tool call needs a human answer.
type Mode string

const (
	// ModeReadOnly refuses mutating tools outright.
	ModeReadOnly Mode = "read-only"
	// ModeAsk prompts before each mutating call. The default.
	ModeAsk Mode = "ask"
	// ModeAcceptEdits auto-approves file edits but still gates bash.
	ModeAcceptEdits Mode = "accept-edits"
	// ModeYolo never prompts. Only for throwaway sandboxes.
	ModeYolo Mode = "yolo"
)

// Modes lists every valid mode, for flag validation and help text.
var Modes = []Mode{ModeReadOnly, ModeAsk, ModeAcceptEdits, ModeYolo}

// ValidMode reports whether s names a real mode.
func ValidMode(s string) bool {
	for _, m := range Modes {
		if Mode(s) == m {
			return true
		}
	}
	return false
}

// Decision is the outcome of one approval check.
type Decision struct {
	Allowed bool
	Reason  string
}

// Gate decides whether a tool call may run.
type Gate struct {
	Mode        Mode
	Interactive bool
	// remembered holds tool names the user answered "always" for, this session only.
	remembered map[string]bool

	// The reader is kept on the gate rather than rebuilt per prompt: a fresh
	// bufio.Reader would read ahead and swallow the next prompt's input.
	in  *bufio.Reader // swappable for tests
	out io.Writer
}

// NewGate builds a gate reading approvals from in and prompting on out.
func NewGate(mode Mode, interactive bool, in io.Reader, out io.Writer) *Gate {
	g := &Gate{Mode: mode, Interactive: interactive, remembered: map[string]bool{}, out: out}
	if in != nil {
		g.in = bufio.NewReader(in)
	}
	return g
}

// Forget clears the "always allow" answers, e.g. after a mode change.
func (g *Gate) Forget() { g.remembered = map[string]bool{} }

// Check gates one tool call. Read-only tools never prompt.
func (g *Gate) Check(toolName string, mutating bool, preview string) Decision {
	if !mutating {
		return Decision{Allowed: true}
	}
	if g.Mode == ModeYolo || g.remembered[toolName] {
		return Decision{Allowed: true}
	}
	if g.Mode == ModeReadOnly {
		return Decision{Reason: "read-only mode: mutating tools are disabled"}
	}
	if g.Mode == ModeAcceptEdits && toolName != "bash" {
		return Decision{Allowed: true}
	}
	if !g.Interactive || g.in == nil {
		return Decision{Reason: "no TTY available to approve this call"}
	}
	return g.prompt(toolName, preview)
}

func (g *Gate) prompt(toolName, preview string) Decision {
	fmt.Fprintf(g.out, "\n\x1b[33m┌ 승인 요청: %s\x1b[0m\n", toolName)
	for i, line := range strings.Split(preview, "\n") {
		if i >= 40 {
			break
		}
		fmt.Fprintf(g.out, "\x1b[33m│\x1b[0m %s\n", line)
	}
	fmt.Fprint(g.out, "\x1b[33m└ [y] 허용  [a] 이 툴은 항상 허용  [n] 거부\x1b[0m\n  > ")

	line, err := g.in.ReadString('\n')
	if err != nil && line == "" {
		return Decision{Reason: "user aborted the approval prompt"}
	}
	switch strings.ToLower(strings.TrimSpace(line)) {
	case "a", "always":
		g.remembered[toolName] = true
		return Decision{Allowed: true}
	case "", "y", "yes":
		return Decision{Allowed: true}
	default:
		return Decision{Reason: "user denied this tool call"}
	}
}

File: deepseek-harness/typescript/src/permissions.ts
/**
 * Approval gate.
 *
 * Every mutating tool call goes through `check` before it runs. Read-only tools
 * never prompt; anything that writes to disk or executes a command does, unless
 * the mode or a remembered answer says otherwise.
 */

import { createInterface } from "node:readline/promises";

// A union of string literals rather than an enum: enums are not erasable
// syntax, and this file has to run under Node's native type stripping.
export const MODES = ["read-only", "ask", "accept-edits", "yolo"] as const;
export type Mode = (typeof MODES)[number];

export interface Decision {
  allowed: boolean;
  reason?: string;
}

export class PermissionGate {
  mode: Mode;
  interactive: boolean;
  /** Tool names the user answered "always" for -- this session only. */
  readonly remembered = new Set<string>();

  constructor(mode: Mode = "ask", interactive = process.stdin.isTTY === true) {
    this.mode = mode;
    this.interactive = interactive;
  }

  async check(toolName: string, mutating: boolean, preview: string): Promise<Decision> {
    if (!mutating) return { allowed: true };
    if (this.mode === "yolo" || this.remembered.has(toolName)) return { allowed: true };
    if (this.mode === "read-only") {
      return { allowed: false, reason: "read-only mode: mutating tools are disabled" };
    }
    if (this.mode === "accept-edits" && toolName !== "bash") return { allowed: true };
    if (!this.interactive) {
      return { allowed: false, reason: "no TTY available to approve this call" };
    }
    return this.prompt(toolName, preview);
  }

  private async prompt(toolName: string, preview: string): Promise<Decision> {
    process.stderr.write(`\n\x1b[33m┌ 승인 요청: ${toolName}\x1b[0m\n`);
    for (const line of preview.split("\n").slice(0, 40)) {
      process.stderr.write(`\x1b[33m│\x1b[0m ${line}\n`);
    }
    process.stderr.write("\x1b[33m└ [y] 허용  [a] 이 툴은 항상 허용  [n] 거부\x1b[0m\n");

    const rl = createInterface({ input: process.stdin, output: process.stderr });
    try {
      const answer = (await rl.question("  > ")).trim().toLowerCase();
      if (answer === "a" || answer === "always") {
        this.remembered.add(toolName);
        return { allowed: true };
      }
      if (answer === "" || answer === "y" || answer === "yes") return { allowed: true };
      return { allowed: false, reason: "user denied this tool call" };
    } catch {
      return { allowed: false, reason: "user aborted the approval prompt" };
    } finally {
      rl.close();
    }
  }
}
